src/dev/amdgpu/pm4_packet_processor.cc

```diff
@@ -181,11 +181,28 @@ PM4PacketProcessor::decodeNext(PM4Queue *q)
         dmaReadVirt(getGARTAddr(q->rptr()), sizeof(uint32_t), cb,
                     &cb->dmaBuffer);
     } else {
+        // Reached the end of processable data in the queue. Switch out of IB
+        // if this is an indirect buffer.
+        assert(q->rptr() == q->wptr());
         q->processing(false);
         if (q->ib()) {
             q->ib(false);
             decodeNext(q);
         }
+
+        // Write back rptr when the queue is empty. For static queues which
+        // are not unmapped, this is how the driver knows there is enough
+        // space in the queue to continue writing packets to the ring buffer.
+        if (q->getMQD()->aqlRptr) {
```

Comment:
I'm not sure how this check confirms that the queue is empty?

Reply:
It will be empty it rptr == wptr by HSA spec. -- This is the != change above.  Basically, this is just a nullptr check.  The HSA spec basically says "it's up to the device" when to report back the rptr.  As a simulator there isn't much downside to report it as much as possible.

Reply:
Most apps do not see this case. I used the bug label after running an app that uses (abuses?) 64 HSA queues